### DISCUSSION TITLE
How to populate a ui.tree?

### DISCUSSION THREAD

**Discussion by JS-Aibel** [zauberzeug/nicegui#804] [Ideas / Feature Requests]
Is there an easy way to populate a [ui.tree?](https://nicegui.io/documentation/tree)
Adding a node in the library [treelib](https://treelib.readthedocs.io/en/latest/) is done with a one liner: `tree.create_node("Id1", "test", parent="root")`
I feel its cumbersome to build the whole data structure needed for the ui.tree.

**Possible suggestions:**
1.	An easy way to add child nodes like the example above.
2.	A method from pandas dataframe to ui.tree structure. Where the arguments are the id and the parent columns. With optional arguments for additional columns/data.
3.	A method to populate a ui.tree() with a treelib data structure.

**Comment by abulka** [comment]
There's a vaguely related discussion https://github.com/zauberzeug/nicegui/discussions/778#discussioncomment-5647634 about populating a ui.tree() from objects in the hope of getting some sort of reactivity.

There's also a Discord discussion re a possible bug re setting the tree expanded property https://discord.com/channels/1089836369431498784/1097428500639653990/1097428500639653990

Perhaps one day something like this would be possible?
```python
modifyable_tree = ui.tree([
    {'id': 'numbers', 'children': [
        {'id': '1'}, {'id': '2'}]},
    {'id': 'letters', 'children': [{'id': 'A'}, {'id': 'B'}]},
], label_key='id', on_select=lambda e: ui.notify(e.value))

# add button to add a new item to the tree
ui.button('add item', on_click=lambda: modifyable_tree.nodes.append({'id': 'C'})) # <-- doesn't work, how to do this?
```

I don't think `ui.tree` is fully fleshed out at the moment.  Perhaps I don't know how to drive NiceGui properly, but I ended up having to use vue + electron + quasar e.g.:
```js
const simple = ref([
  {
    label: 'Relax Hotel',
    children: [
      {
        label: 'Room view',
        icon: 'photo'
      }
    ]
  }
])
const expanded = ref(['Relax Hotel', 'Room amenities'])

...

<q-btn label="Expand" @click="expanded = ['Relax Hotel']" />
<q-btn label="Expand2" @click="expanded = ['Relax Hotel', 'Room amenities']" />

<q-btn label="Change a node" @click="simple[0].children[1].label = 'FRED'" />

<q-tree
  v-model:expanded="expanded"
  :nodes="simple"
  node-key="label"
  selected-color="primary"
  default-expand-all
  dark
/>
```
Not only is `expanded` nicely bound using v-model, and settable, I can even change tree node data object and the tree magically updates as expected. I hope one day all this is possible via NiceGui.

**Comment by vindevoy** [comment]
I was able to populate the tree using just using a correctly formatted list.  This is the first part of this code.  I supply this to ui.tree() and I get a nice treeview


```py
[
    {'id': '/ald', 'children': [
        {'id': '/ald/wmg', 'children': [
            {'id': '/ald/wmg/dash_kpi.html', 'label': 'Dash KPI'}, 
            {'id': '/ald/wmg/slabcarrier_kpi-copy1.html', 'label': 'SlabCarrier KPI-Copy1'}, 
            {'id': '/ald/wmg/slabcarrier_kpi.html', 'label': 'SlabCarrier KPI'}, 
            {'id': '/ald/wmg/untitled.html', 'label': 'Untitled'}], 'label': 'WMG'}], 'label': 'ALD'}, 
            {'id': '/coo', 'children': [
                {'id': '/coo/coo_reliability_auto.html', 'label': 'coo_reliability_auto'}, 
                {'id': '/coo/kritische_alarmen_extractoren_week.html', 'label': 'Kritische_alarmen_extractoren_week'}, 
                {'id': '/coo/kritische_alarmen_extractoren_year.html', 'label': 'Kritische_alarmen_extractoren_year'}, 
                {'id': '/coo/test_dropdown.html', 'label': 'test_dropdown'}
            ], 'label': 'COO'}, 
            {'id': '/ghv', 'children': [
                {'id': '/ghv/kwaliteit', 'children': [
                    {'id': '/ghv/kwaliteit/nuttige_fracties.html', 'label': 'Nuttige fracties'}
                ], 'label': 'Kwaliteit'}, 
                {'id': '/ghv/logistiek_schepen', 'children': [
                    {'id': '/ghv/logistiek_schepen/aankomsten_schepen_(beax-begne).html', 'label': 'Aankomsten schepen (BEAX-BEGNE)'}, 
                    {'id': '/ghv/logistiek_schepen/aankomsten_schepen_v0.html', 'label': 'Aankomsten schepen V0'}
                ], 'label': 'Logistiek schepen'}, 
```

**Reply by vindevoy** [reply]
![image](https://github.com/zauberzeug/nicegui/assets/26721992/ada84efa-0e76-4df5-a9ce-c2c4e40611ee)

Screenshot of the tree with the code "tree" above in this code


```
 ui.tree(tree, on_select=lambda e: self._open_page_from_menu(router=router, url=e.value)).\
                props(add="dense").classes("-m-2 p-0").style("color: #D0D0D0;").props("no-selection-unset=true")

```

**Reply by sunnymax2002** [reply]
I recently created a simple wrapper over Treelib to solve a similar problem. Now I can initialize the Tree, and dynamically add nodes just by specifying its parent and refresh the tree Check out the [GitHub Gist](https://gist.github.com/sunnymax2002/cf5bf4b85de56eb3fb59537b771caa1f) and share feedback to improve, or possibly add to nicegui

**Reply by flooxo** [reply]
> I recently created a simple wrapper over Treelib to solve a similar problem. Now I can initialize the Tree, and dynamically add nodes just by specifying its parent and refresh the tree Check out the [GitHub Gist](https://gist.github.com/sunnymax2002/cf5bf4b85de56eb3fb59537b771caa1f) and share feedback to improve, or possibly add to nicegui

Thanks! Very helpful wrapper :)

I forked the gist and added the functionality to add descriptions for the nodes
GitHub Gist: [TreeView wrapper with description](https://gist.github.com/flooxo/87e244a5717087c2b22af110874380c1)

**Comment by ravi-2912** [comment]
Is there a way to dynamically add or update tree nodes, maybe through a state?
